posts/MarcoB/2024-07-12-embeddable-inheritance.adoc

```diff
@@ -0,0 +1,240 @@
+====
+[source, sql, indent=0]
+----
+    insert
+    into
+        Owner
+        (age, fins, mother, name, animal_type, id)
+    values
+        (3, null, 'Gatta', 'Arial', 'C', 1)
+----
+====
+
+[[type-treat]]
+== Support for `type()` and `treat()` operators
+
+Of course, the `type()` and `treat()` functions are also supported for embeddable inheritance and can serve to explicitly refer to the type information of `@Embedded` properties in queries.
+
+[[hql-function-type]]
+===== `type()`
+
+The function `type()` evaluates to the concrete type, that is, the Java `Class`, of the referenced entity or embeddable. This is useful when dealing with polymorphic embeddables:
+====
+[source, java, indent=0]
+----
+// Retrieve the type of an embeddable property
+Class<?> petType = entityManager.createQuery(
+	"select type(o.pet) " +
+	"from Owner o " +
+	"where o.id = 1",
+	Class.class)
+.getSingleResult();
+
+// Restrict the embeddable property to a specific subtype
+List<Owner> catOwners = entityManager.createQuery(
+	"select o " +
+	"from Owner o " +
+	"where type(o.pet) = Cat",
+	Owner.class)
+.getResultList();
+----
+====
+
+[[hql-function-treat]]
+===== `treat()`
+
+The function `treat()` may be used to narrow the type of an identification variable:
+====
+[source, java, indent=0]
+----
+List<Owner> owners = entityManager.createQuery(
+	"select o " +
+	"from Owner o " +
+	"where treat(o.pet as Cat).mother = :mother",
+	Owner.class)
+.setParameter( "mother", mother )
+.getResultList();
+----
+====
+
+Once again, see the types and typecasts link:https://docs.jboss.org/hibernate/orm/current/userguide/html_single/Hibernate_User_Guide.html#hql-functions-typecasts[user guide chapter] for more details.
+
+[[benefits-and-limitations]]
+== Benefits and Limitations
+
+*Benefits*:
+
+Finally, I've listed here a couple benefits and limitations that characterize this new feature:
+
+* *Code reusability*: Common fields are defined in the parent embeddable classes, promoting reuse and reducing redundancy.
+* *Polymorphic queries*: You can use the `type()` and `treat()` functions to handle polymorphic queries effectively.
+
+*Limitations*:
+
+* *Not supported for composite ids*: inheritance is not supported for embeddables used as primary keys.
+* *Complexity*: Managing discriminator columns can add complexity to your database schema, especially when dealing with multiple embeddable properties.
+
+[[outlooks]]
+== Outlooks
+
+Embeddable inheritance in Hibernate provides an additional tool for designing clean, modular, and reusable data models. By leveraging discriminator columns, our framework allows you to map complex inheritance hierarchies to relational database structures seamlessly. While it adds a layer of complexity, the benefits in terms of code maintainability and query capabilities often outweigh the downsides.
```

Comment:
```suggestion
Embeddable inheritance in Hibernate ORM provides an additional tool for designing clean, modular, and reusable data models. By leveraging discriminator columns, our framework allows you to map complex inheritance hierarchies to relational database structures seamlessly. While it adds a layer of complexity, the benefits in terms of code maintainability and query capabilities often outweigh the downsides.
```
🙈 😈

Reply:
Thank you so much for all the suggestions!